ethcore/src/engines/clique/mod.rs

```diff
@@ -198,13 +198,15 @@ impl Clique {
 			step_service: None,
 		};
 
-		let res = Arc::new(engine);
+		let mut engine= Arc::new(engine);
 
 		if our_params.period > 0 {
-			engine.step_service = Some(StepService::start(Arc::downgrade(&res) as Weak<Engine<_>>));
+			let engine_weak = Arc::downgrade(&engine.clone());
+			let step_service = StepService::start(engine_weak as Weak<Engine<_>>);
+			Arc::get_mut(&mut engine).map(|e| e.step_service = Some(step_service));
```

Comment:
I don't think this will work because we got two references to the same value, https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut

Reply:
Well, it compiles with no warning and runs (seemingly) ok! It feels iffy for sure though.

Do you think it works because of the 5sec delay in the thread started by `StepService`?

Reply:
Hmm, indeed this shouldn't work - can we replace `map` with `expect` to make sure it's always there (cause that's what we assume).
Maybe let's separate creation of `StepService` from actually starting the inner thread? So that we can do:
```
let mut engine = ...;
let step_service = StepService::new();
engine.step_service = step_service;
let engine = Arc::new(engine);
let weak = Arc::downgrade(&engine);
engine.step_service.start(weak);
```

Reply:
@tomusdrw excellent idea, I'll see where it takes me. Thanks!

Reply:
`engine.step_service.start(weak);` does not work as long as `step_service.start()` takes `&mut self` (can't move out of an `Arc`) so need some tinkering. Hmm.

Reply:
@dvdplm because you can't store the thread handle? Maybe put it to `Mutex`, since it's not hot code anyway, and we just need to store it.

Reply:
> because you can't store the thread handle?

Precisely. Using a `Mutex` most likely works; possibly reducing `StepService` to an `Arc<AtomicBool>` (i.e. not storing the thread handle at all) and let the thread run wild and free could work.

I am however pausing work on this for the moment because shutdown is currently broken elsewhere and I need to fix that before coming back to this. :/

Reply:
Ok, it could potentially be this https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/engines/clique/mod.rs#L749? 

I expect it `Client` to aliased there too, have you seen the `warn` print?

Reply:
@niklasad1 with [the other shutdown bug](https://github.com/paritytech/parity-ethereum/pull/10689) I do see it. Without it, that code is never called.